File: src/fg_agent_id/address.py
"""Self-certifying agent addresses: ``amp:key:<base58(ed25519-signing-pub)>``.

The address *is* the signing public key, so any signed artifact from an agent
can be verified with nothing but its address — no registry required.
"""

from __future__ import annotations

from .errors import AddressError
from .keys import base58_decode, base58_encode

ADDRESS_PREFIX = "amp:key:"


def address_from_signing_key(signing_public: bytes) -> str:
    if len(signing_public) != 32:
        raise AddressError("signing public key must be 32 raw bytes")
    return ADDRESS_PREFIX + base58_encode(signing_public)


def signing_key_from_address(address: str) -> bytes:
    if not address.startswith(ADDRESS_PREFIX):
        raise AddressError(f"not an AMP address: {address!r}")
    try:
        key = base58_decode(address[len(ADDRESS_PREFIX):])
    except ValueError as exc:
        raise AddressError(f"invalid base58 in address: {address!r}") from exc
    if len(key) != 32:
        raise AddressError(f"address does not decode to a 32-byte key: {address!r}")
    return key

File: js/src/address.ts
/** Self-certifying agent addresses: `amp:key:<base58(ed25519-signing-pub)>`. */

import { base58Decode, base58Encode } from "./base58.js";
import { AddressError } from "./errors.js";

export const ADDRESS_PREFIX = "amp:key:";

export function addressFromSigningKey(signingPublic: Uint8Array): string {
  if (signingPublic.length !== 32) {
    throw new AddressError("signing public key must be 32 raw bytes");
  }
  return ADDRESS_PREFIX + base58Encode(signingPublic);
}

export function signingKeyFromAddress(address: string): Uint8Array {
  if (!address.startsWith(ADDRESS_PREFIX)) {
    throw new AddressError(`not an AMP address: ${JSON.stringify(address)}`);
  }
  let key: Uint8Array;
  try {
    key = base58Decode(address.slice(ADDRESS_PREFIX.length));
  } catch {
    throw new AddressError(`invalid base58 in address: ${JSON.stringify(address)}`);
  }
  if (key.length !== 32) {
    throw new AddressError(`address does not decode to a 32-byte key: ${JSON.stringify(address)}`);
  }
  return key;
}
